pkg/server/plugin/nodeattestor/devid/devid.go

```diff
@@ -0,0 +1,487 @@
+	}
+
+	// Unmarshall received attestation data
+	attData := new(common_devid.AttestationRequest)
+	err = json.Unmarshal(req.AttestationData.Data, attData)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "unable to unmarshall attestation data: %v", err)
+	}
+
+	// Decode attestation data
+	if len(attData.DevIDCert) == 0 {
+		return status.Error(codes.InvalidArgument, "no DevID certificate to attest")
+	}
+
+	devIDCert, err := x509.ParseCertificate(attData.DevIDCert)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "unable to parse DevID certificate: %v", err)
+	}
+
+	// Verify DevID certificate chain of trust
+	err = verifyDevIDSignature(devIDCert, conf.devIDRoots)
+	if err != nil {
+		return status.Errorf(codes.Unauthenticated, "unable to verify DevID signature: %v", err)
+	}
+
+	// Issue a DevID challenge (to prove the possession of the DevID private key).
+	devIDChallenge, err := newDevIDChallenge()
+	if err != nil {
+		return status.Errorf(codes.Internal, "unable to generate challenge: %v", err)
+	}
+
+	// Verify DevID residency (if configured)
+	var nonce []byte
+	var credActivationChallenge *common_devid.CredActivation
+	if conf.checkDevIDResidency {
+		credActivationChallenge, nonce, err = verifyDevIDResidency(attData, conf.ekRoots)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Marshal challenges
+	challenge, err := json.Marshal(common_devid.ChallengeRequest{
+		DevID:          devIDChallenge,
+		CredActivation: credActivationChallenge,
+	})
+	if err != nil {
+		return status.Errorf(codes.Internal, "unable to marshal challenges data: %v", err)
+	}
+
+	// Send challenges to the agent
+	err = stream.Send(&nodeattestorv0.AttestResponse{Challenge: challenge})
+	if err != nil {
+		return status.Errorf(status.Code(err), "unable to send challenges: %v", err)
+	}
+
+	// Receive challenges response
+	responseReq, err := stream.Recv()
+	if err != nil {
+		return status.Errorf(status.Code(err), "unable to receive challenges response: %v", err)
+	}
+
+	// Unmarshal challenges response
+	challengeResponse := &common_devid.ChallengeResponse{}
+	if err = json.Unmarshal(responseReq.Response, challengeResponse); err != nil {
+		return status.Errorf(codes.InvalidArgument, "unable to unmarshall challenges response: %v", err)
+	}
+
+	// Verify DevID challenge
+	err = verifyDevIDChallenge(devIDCert, devIDChallenge, challengeResponse.DevID)
+	if err != nil {
+		return status.Errorf(codes.Unauthenticated, "devID challenge verification failed: %v", err)
+	}
+
+	// Verify credential activation challenge (if configured)
+	if conf.checkDevIDResidency {
+		err = verifyCredActivationChallenge(nonce, challengeResponse.CredActivation)
+		if err != nil {
+			return status.Errorf(codes.Unauthenticated, "credential activation failed: %v", err)
+		}
+	}
+
+	// Create SPIFFE ID and selectors
+	certSelectors := selectorsFromCertificate(common_devid.PluginName, "certificate", devIDCert)
+	fingerprint := x509pop.Fingerprint(devIDCert)
```

Comment:
We should probably not use this function directly, since it won't be obvious that there are consumers outside of the `x509pop` plugin if we ever decide to change it.

Moving the function to `x509utils` package seems like a good path forward.. @azdagron do you have any preference on where something like this should live?

Reply:
It's a tiny function and changing it will impact the generated SPIFFE IDs for agents, so I _think_ we should probably just duplicate it? I don't see us ever changing it, but if we do, it should be with great intent. If we stuffed it in something common like `x509util`, I think the probability of it being inadvertently changed goes up.

Reply:
Sounds good, I'll duplicate it.